src/Schema/AST/ASTHelper.php

```diff
@@ -379,4 +390,64 @@ public static function internalFieldName(FieldDefinitionNode $field): string
             ? static::directiveArgValue($renameDirectiveNode, 'attribute')
             : $field->name->value;
     }
+
+    /**
+     * Adds a directive to a node.
+     *
+     * @api
+     */
+    public static function addDirectiveToNode(string $directiveSource, ScalarTypeDefinitionNode|ScalarTypeExtensionNode|ObjectTypeDefinitionNode|ObjectTypeExtensionNode|InterfaceTypeDefinitionNode|InterfaceTypeExtensionNode|UnionTypeDefinitionNode|UnionTypeExtensionNode|EnumTypeDefinitionNode|EnumTypeExtensionNode|InputObjectTypeDefinitionNode|InputObjectTypeExtensionNode|FieldDefinitionNode|InputValueDefinitionNode|EnumValueDefinitionNode $node): \Nuwave\Lighthouse\Support\Contracts\Directive
+    {
+        $directiveNode = Parser::directive($directiveSource);
+        $node->directives[] = $directiveNode;
+
+        $directiveLocator = Container::getInstance()->make(DirectiveLocator::class);
+        $directiveInstance = $directiveLocator->create($directiveNode->name->value);
+        if ($directiveInstance instanceof BaseDirective) {
+            $directiveInstance->hydrate($directiveNode, $node);
+        }
+
+        return $directiveInstance;
+    }
+
+    /**
+     * Add a directive to a field definition dynamically.
+     *
+     * @api
+     */
+    public static function addDirectiveToFieldDefinition(
+        string $directiveSource,
+        DocumentAST &$documentAST,
+        FieldDefinitionNode &$fieldDefinition,
+        ObjectTypeDefinitionNode|InterfaceTypeDefinitionNode &$parentType,
+    ): void {
+        $directiveInstance = self::addDirectiveToNode($directiveSource, $fieldDefinition);
+
+        if (!($directiveInstance instanceof FieldManipulator)) {
+            throw new DefinitionException("The dynamically added directive [{$directiveSource}] must implement the FieldManipulator interface.");
+        }
+
+        $directiveInstance->manipulateFieldDefinition($documentAST, $fieldDefinition, $parentType);
+    }
+
+    /**
+     * Add a directive to an arg definition dynamically.
+     *
+     * @api
+     */
+    public static function addDirectiveToArgDefinition(
+        string $directiveSource,
+        DocumentAST &$documentAST,
+        InputValueDefinitionNode &$argDefinition,
+        FieldDefinitionNode &$parentField,
+        ObjectTypeDefinitionNode|InterfaceTypeDefinitionNode &$parentType,
+    ): void {
+        $directiveInstance = self::addDirectiveToNode($directiveSource, $argDefinition);
+
+        if (!($directiveInstance instanceof ArgManipulator)) {
+            throw new DefinitionException("The dynamically added directive [{$directiveSource}] must implement the ArgManipulator interface.");
+        }
+
+        $directiveInstance->manipulateArgDefinition($documentAST, $argDefinition, $parentField, $parentType);
+    }
```

Comment:
I would prefer those functions to be removed and done ad-hoc instead.

Reply:
Sure, I can remove them. Would be open to make `ASTHelper` macroable then? Then it wouldn't be necessary to create an own class and keep everything there.

Reply:
I don't see a compelling enough technical reason for a class like `ASTHelper` that is basically just a collection of static functions - which in an ideal world would just be plain old PHP functions. If it had internal state that macros needed access to, I could see it being useful.

Reply:
Okay, fine for me. It's removed 😊